File: operator/controllers/acls.go
package controllers

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"sigs.k8s.io/controller-runtime/pkg/client"

	v1alpha1 "github.com/woestebanaan/skafka/operator/api/v1alpha1"
)

// ACLFile is the in-memory representation of __cluster/acls.json.
// Schema is deliberately UNCHANGED by gh #135 — the broker still reads
// this exact layout from disk. Only the CR-side authoring story changed:
// pre-gh #135 the operator merged a separate KafkaAcl CR + a
// KafkaUserGroup expansion into entries; post-gh #135 it iterates
// KafkaUser CRs and pulls the inline `spec.authorization.acls` slice.
// The on-disk format ("Allow"/"Deny" capitalisation, principal prefix)
// matches what the broker's loader already expects, so no broker-side
// code or data migration is needed.
type ACLFile struct {
	Version int        `json:"version"`
	ACLs    []ACLEntry `json:"acls"`
}

// ACLEntry is one access control rule as stored on disk.
type ACLEntry struct {
	Principal  string      `json:"principal"`
	Resource   ACLResource `json:"resource"`
	Operations []string    `json:"operations"`
	Permission string      `json:"permission"`
}

// ACLResource mirrors the in-CR resource shape for JSON output.
type ACLResource struct {
	Type        string `json:"type"`
	Name        string `json:"name"`
	PatternType string `json:"patternType"`
}

func aclsPath(dataDir string) string {
	return filepath.Join(dataDir, "__cluster", "acls.json")
}

// reconcileACLs lists all KafkaUser objects in namespace, collects their
// inline `spec.authorization.acls` rules, and atomically writes the
// merged set to dataDir/__cluster/acls.json.
//
// Called from KafkaUserReconciler on every KafkaUser change so a single
// user-level edit is reflected on disk within one reconcile cycle. Safe
// to call from concurrent reconciles — the atomic rename guarantees the
// broker only ever sees a complete file.
func reconcileACLs(ctx context.Context, c client.Client, namespace, dataDir string) error {
	var userList v1alpha1.KafkaUserList
	if err := c.List(ctx, &userList, client.InNamespace(namespace)); err != nil {
		return err
	}

	var entries []ACLEntry
	for _, u := range userList.Items {
		if u.DeletionTimestamp != nil {
			continue
		}
		if u.Spec.Authorization == nil || len(u.Spec.Authorization.ACLs) == 0 {
			continue
		}
		principal := "User:" + u.Name
		for _, acl := range u.Spec.Authorization.ACLs {
			entries = append(entries, aclToEntry(principal, acl))
		}
	}

	if err := os.MkdirAll(filepath.Join(dataDir, "__cluster"), 0o775); err != nil {
		return err
	}
	return writeAtomic(aclsPath(dataDir), &ACLFile{Version: 1, ACLs: entries})
}

// aclToEntry projects one KafkaUserACL (Strimzi-style, with
// `type: allow|deny` lowercased) onto the on-disk ACLEntry format
// (`permission: Allow|Deny` capitalised). Defaults the pattern type
// to "literal" when the CR didn't set one and the type to "allow"
// when omitted — both mirror Strimzi defaults.
func aclToEntry(principal string, acl v1alpha1.KafkaUserACL) ACLEntry {
	patternType := acl.Resource.PatternType
	if patternType == "" {
		patternType = "literal"
	}
	aclType := acl.Type
	if aclType == "" {
		aclType = "allow"
	}
	// On-disk format is the historical "Allow" / "Deny" capitalisation
	// (broker AclEngine matches case-sensitively). Translate at the
	// boundary so the public CR stays Strimzi-faithful.
	permission := "Allow"
	if strings.EqualFold(aclType, "deny") {
		permission = "Deny"
	}
	return ACLEntry{
		Principal: principal,
		Resource: ACLResource{
			Type:        acl.Resource.Type,
			Name:        acl.Resource.Name,
			PatternType: patternType,
		},
		Operations: acl.Operations,
		Permission: permission,
	}
}

// writeAtomic serialises v to JSON and writes it to path via a tmp+rename.
func writeAtomic(path string, v any) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0640); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

File: crates/sk-operator-controllers/src/acls.rs
//! Walks every `KafkaUser` CR in the operator's namespace, projects
//! their inline `spec.authorization.acls` into the on-disk
//! `acls.json` shape, and atomic-writes the file under
//! `/data/__cluster/`. Same JSON shape the broker's `AclEngine` (in
//! `sk-auth`) already consumes — no broker-side code change.
//!
//! `acl_to_entry` is the per-rule translation step; it's `pub`
//! because workstream G's tests assert the defaulting behaviour
//! (gh #137: pattern_type=literal, type=allow when CR field empty).
//!
//! Note the historical capitalisation of `permission`: the on-disk
//! field is `Allow` / `Deny` (capital initial) while the CR field is
//! `allow` / `deny` (lowercase, Strimzi-faithful). We translate at
//! the boundary so the public surface stays Strimzi-shape while the
//! broker side keeps the existing case-sensitive matcher.

use std::path::{Path, PathBuf};

use kube::{api::ListParams, Api, Client};
use serde::{Deserialize, Serialize};
use sk_operator_api::{KafkaUser, KafkaUserAcl};

use crate::errors::ControllerError;

/// On-disk shape of `__cluster/acls.json`.
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AclFile {
    pub version: u32,
    #[serde(default)]
    pub acls: Vec<AclEntry>,
}

/// One ACL rule as stored on disk. Field shapes are pinned
/// byte-for-byte (the broker's existing `sk-auth` loader
/// gates on these names).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AclEntry {
    /// `User:<name>` — note the literal `User:` prefix the broker
    /// matches against.
    pub principal: String,

    pub resource: AclResource,

    pub operations: Vec<String>,

    /// `Allow` or `Deny` (capitalised). Case-sensitive on the broker
    /// side — see `crates/sk-auth/src/acls.rs`.
    pub permission: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AclResource {
    #[serde(rename = "type")]
    pub kind: String,
    pub name: String,
    pub pattern_type: String,
}

/// `<data_dir>/__cluster/acls.json`.
pub fn acls_path(data_dir: &Path) -> PathBuf {
    data_dir.join("__cluster").join("acls.json")
}

/// List every `KafkaUser` CR in `namespace` and rebuild `acls.json`
/// from the union of their inline `spec.authorization.acls`.
///
/// Called from `KafkaUserReconciler` on every CR change so a single
/// user-level edit propagates within one reconcile cycle. Safe to
/// call from concurrent reconciles — the underlying `atomic_write`
/// guarantees the broker only ever sees a complete file.
///
/// Users with `metadata.deletionTimestamp` set are excluded from the
/// rebuild (their rules go away as soon as the delete is observed,
/// without waiting for the final `Deleted` event).
pub async fn reconcile_acls(
    client: &Client,
    namespace: &str,
    data_dir: &Path,
) -> Result<(), ControllerError> {
    let api: Api<KafkaUser> = Api::namespaced(client.clone(), namespace);
    let users = api.list(&ListParams::default()).await?;

    let mut entries: Vec<AclEntry> = Vec::new();
    for u in &users.items {
        if u.metadata.deletion_timestamp.is_some() {
            continue;
        }
        let Some(authz) = u.spec.authorization.as_ref() else {
            continue;
        };
        if authz.acls.is_empty() {
            continue;
        }
        let Some(name) = u.metadata.name.as_deref() else {
            continue;
        };
        let principal = format!("User:{name}");
        for acl in &authz.acls {
            entries.push(acl_to_entry(&principal, acl));
        }
    }

    write_acl_file(
        data_dir,
        &AclFile {
            version: 1,
            acls: entries,
        },
    )
}

/// Project one `KafkaUserACL` (Strimzi-style, lowercased
/// `type: allow|deny`) onto the on-disk shape (capitalised
/// `permission: Allow|Deny`). Defaults the pattern_type to `literal`
/// when empty and the type to `allow` when empty — both mirror
/// Strimzi defaults per gh #137 (no apiserver defaulting).
pub fn acl_to_entry(principal: &str, acl: &KafkaUserAcl) -> AclEntry {
    let pattern_type = if acl.resource.pattern_type.is_empty() {
        "literal"
    } else {
        acl.resource.pattern_type.as_str()
    };
    let permission = if acl.kind.eq_ignore_ascii_case("deny") {
        "Deny"
    } else {
        "Allow"
    };
    AclEntry {
        principal: principal.to_string(),
        resource: AclResource {
            kind: acl.resource.kind.clone(),
            name: acl.resource.name.clone(),
            pattern_type: pattern_type.to_string(),
        },
        operations: acl.operations.clone(),
        permission: permission.to_string(),
    }
}

/// Read `acls.json` (returns empty `{version: 1, acls: []}` when
/// the file is absent). Useful for tests + sweep-style assertions.
pub fn read_acl_file(data_dir: &Path) -> Result<AclFile, ControllerError> {
    let path = acls_path(data_dir);
    match std::fs::read(&path) {
        Ok(bytes) => Ok(serde_json::from_slice(&bytes)?),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(AclFile {
            version: 1,
            acls: Vec::new(),
        }),
        Err(e) => Err(ControllerError::Io(e)),
    }
}

/// Atomic write of `acls.json`. Stamps `version = 1`.
pub fn write_acl_file(data_dir: &Path, file: &AclFile) -> Result<(), ControllerError> {
    let cluster_dir = data_dir.join("__cluster");
    std::fs::create_dir_all(&cluster_dir)?;
    let mut stamped = file.clone();
    stamped.version = 1;
    let fs = sk_storage::fs::RealFs::new();
    sk_storage::atomic_write::atomic_write_json_pretty(&fs, &cluster_dir, "acls.json", &stamped)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use sk_operator_api::KafkaUserAclResource;

    fn rule(kind: &str, res_kind: &str, name: &str, pattern: &str, ops: &[&str]) -> KafkaUserAcl {
        KafkaUserAcl {
            resource: KafkaUserAclResource {
                kind: res_kind.into(),
                name: name.into(),
                pattern_type: pattern.into(),
            },
            operations: ops.iter().map(|s| (*s).to_string()).collect(),
            kind: kind.into(),
            host: String::new(),
        }
    }

    #[test]
    fn acl_to_entry_defaults_pattern_to_literal_when_empty() {
        let e = acl_to_entry(
            "User:alice",
            &rule("allow", "topic", "events", "", &["Read"]),
        );
        assert_eq!(e.resource.pattern_type, "literal");
    }

    #[test]
    fn acl_to_entry_capitalises_permission() {
        let allow = acl_to_entry("User:a", &rule("allow", "topic", "x", "literal", &["Read"]));
        let deny = acl_to_entry("User:a", &rule("deny", "topic", "x", "literal", &["Read"]));
        let empty = acl_to_entry("User:a", &rule("", "topic", "x", "literal", &["Read"]));
        assert_eq!(allow.permission, "Allow");
        assert_eq!(deny.permission, "Deny");
        // Empty kind defaults to Allow (gh #137 — operator-side default).
        assert_eq!(empty.permission, "Allow");
    }

    #[test]
    fn acl_to_entry_passes_through_pattern_when_set() {
        let e = acl_to_entry("User:a", &rule("allow", "topic", "x", "prefix", &["Read"]));
        assert_eq!(e.resource.pattern_type, "prefix");
    }

    #[test]
    fn acl_file_roundtrip_via_disk() {
        let tmp = tempfile::tempdir().unwrap();
        let f = AclFile {
            version: 1,
            acls: vec![AclEntry {
                principal: "User:alice".into(),
                resource: AclResource {
                    kind: "topic".into(),
                    name: "events".into(),
                    pattern_type: "literal".into(),
                },
                operations: vec!["Read".into(), "Describe".into()],
                permission: "Allow".into(),
            }],
        };
        write_acl_file(tmp.path(), &f).unwrap();
        let back = read_acl_file(tmp.path()).unwrap();
        assert_eq!(back, f);
    }

    #[test]
    fn read_acl_file_absent_returns_empty() {
        let tmp = tempfile::tempdir().unwrap();
        let f = read_acl_file(tmp.path()).unwrap();
        assert_eq!(f.version, 1);
        assert!(f.acls.is_empty());
    }
}
